file/shard.go

```diff
@@ -15,56 +16,149 @@ type shardNodeFile struct {
-	done      bool
 	rdr       io.Reader
+	offset    int64
 }
 
 var _ ipld.Node = (*shardNodeFile)(nil)
 
-func (s *shardNodeFile) Read(p []byte) (int, error) {
-	if s.done {
-		return 0, io.EOF
+func (s *shardNodeFile) makeReader() (io.Reader, error) {
+	links, err := s.substrate.LookupByString("Links")
+	if err != nil {
+		return nil, err
 	}
-	// collect the sub-nodes on first use
-	if s.rdr == nil {
-		links, err := s.substrate.LookupByString("Links")
+	readers := make([]io.Reader, 0)
+	lnki := links.ListIterator()
+	at := int64(0)
+	for !lnki.Done() {
+		_, lnk, err := lnki.Next()
 		if err != nil {
-			return 0, err
+			return nil, err
+		}
+		sz, err := lnk.LookupByString("Tsize")
+		if err != nil {
+			return nil, err
+		}
+		childSize, err := sz.AsInt()
+		if err != nil {
+			return nil, err
+		}
+		if s.offset > at+childSize {
+			continue
+		}
+		lnkhash, err := lnk.LookupByString("Hash")
+		if err != nil {
+			return nil, err
+		}
+		lnklnk, err := lnkhash.AsLink()
+		if err != nil {
+			return nil, err
+		}
+		// todo: defer load until first read.
+		target, err := s.lsys.Load(ipld.LinkContext{Ctx: s.ctx}, lnklnk, protoFor(lnklnk))
+		if err != nil {
+			return nil, err
 		}
-		readers := make([]io.Reader, 0)
-		lnki := links.ListIterator()
-		for !lnki.Done() {
-			_, lnk, err := lnki.Next()
-			if err != nil {
-				return 0, err
-			}
-			lnkhash, err := lnk.LookupByString("Hash")
-			if err != nil {
-				return 0, err
-			}
-			lnklnk, err := lnkhash.AsLink()
-			if err != nil {
-				return 0, err
-			}
-			target, err := s.lsys.Load(ipld.LinkContext{Ctx: s.ctx}, lnklnk, protoFor(lnklnk))
-			if err != nil {
-				return 0, err
-			}
 
-			asFSNode, err := NewUnixFSFile(s.ctx, target, s.lsys)
+		asFSNode, err := NewUnixFSFile(s.ctx, target, s.lsys)
+		if err != nil {
+			return nil, err
+		}
+		// fastforward the first one if needed.
+		if at < s.offset {
+			_, err := asFSNode.Seek(s.offset-at, io.SeekStart)
 			if err != nil {
-				return 0, err
+				return nil, err
 			}
-			readers = append(readers, asFSNode)
 		}
-		s.rdr = io.MultiReader(readers...)
+		at += childSize
+		readers = append(readers, asFSNode)
+	}
+	if len(readers) == 0 {
+		return nil, io.EOF
+	}
+	return io.MultiReader(readers...), nil
+}
+
+func (s *shardNodeFile) Read(p []byte) (int, error) {
+	// build reader
+	if s.rdr == nil {
+		rdr, err := s.makeReader()
+		if err != nil {
+			return 0, err
+		}
+		s.rdr = rdr
 	}
 	n, err := s.rdr.Read(p)
-	if err == io.EOF {
+	return n, err
+}
+
+func (s *shardNodeFile) Seek(offset int64, whence int) (int64, error) {
+	if s.rdr != nil {
 		s.rdr = nil
-		s.done = true
 	}
-	return n, err
+	switch whence {
+	case io.SeekStart:
+		s.offset = offset
+	case io.SeekCurrent:
+		s.offset += offset
+	case io.SeekEnd:
+		s.offset = s.length() + offset
+	}
+	return s.offset, nil
+}
+
+func (s *shardNodeFile) length() int64 {
+	// see if we have size specified in the unixfs data. errors fall back to length from links
```

Comment:
I'm slightly worried that, if a node is neither a "data" node nor a "links" node, you'll happily fall back to `return 0` - you won't return any error at all, and you might swallow genuinely useful errors such as those from `DecodeUnixFSData`.